### ISSUE TITLE
defer 链表如何被遍历 | qcrao

### ISSUE THREAD

**Issue by qcrao** [qcrao/qcrao.github.io#57]
https://qcrao.com/2020/03/23/how-to-traverse-defer-links/ 

 日拱一卒，终有所成！

**Comment by Hawken94** [comment]
"如果此参数**不是**指针或引用类型的话，会产生一些意料之外的 bug。" 应该是 "是指针或引用类型"吧

**Comment by qcrao** [comment]
> @Hawken94
> "如果此参数**不是**指针或引用类型的话，会产生一些意料之外的 bug。" 应该是 "是指针或引用类型"吧

比如 defer 里判断 err 是否为 nil，然后做一些操作。写 defer 语句的时候 err 还是 nil，但中间经过一系列操作，最后 err 已经不是 nil 了，这就出问题了。但如果用 *err 就没事……

**Comment by Hawken94** [comment]
> > @Hawken94
> > "如果此参数**不是**指针或引用类型的话，会产生一些意料之外的 bug。" 应该是 "是指针或引用类型"吧
> 
> 比如 defer 里判断 err 是否为 nil，然后做一些操作。写 defer 语句的时候 err 还是 nil，但中间经过一系列操作，最后 err 已经不是 nil 了，这就出问题了。但如果用 *err 就没事……

哦哦，理解了，非常感谢。如果参数是指针类型，是指下面的这个例子吗
``` go
func main() {
	var er error = nil
	err := &er // err is a pointer

	defer func(err1 error) {
		fmt.Println("defer:", err1)
	}(*err)

	*err = errors.New("error")
	fmt.Println("main:", *err)

	// main: error
	// defer: <nil>
}
```

**Comment by qcrao** [comment]
> @Hawken94
> > > @Hawken94
> > > "如果此参数**不是**指针或引用类型的话，会产生一些意料之外的 bug。" 应该是 "是指针或引用类型"吧
> > 
> > 比如 defer 里判断 err 是否为 nil，然后做一些操作。写 defer 语句的时候 err 还是 nil，但中间经过一系列操作，最后 err 已经不是 nil 了，这就出问题了。但如果用 *err 就没事……
> 
> 哦哦，理解了，非常感谢。如果参数是指针类型，是指下面的这个例子吗
> ``` go
> func main() {
> 	var er error = nil
> 	err := &er // err is a pointer
> 
> 	defer func(err1 error) {
> 		fmt.Println("defer:", err1)
> 	}(*err)
> 
> 	*err = errors.New("error")
> 	fmt.Println("main:", *err)
> 
> 	// main: error
> 	// defer: <nil>
> }
> ```


不是。你这其实还是有问题，defer 最后执行的时候，err1 的 nil 其实在 deferproc 的时候就已经确定值了。

理解下这个例子：

```golang
package main

import (
	"errors"
	"fmt"
)

func main() {
	var er error = nil
	err := &er // err is a pointer

	defer func(err1 *error) {
		fmt.Println("defer:", *err1)
	}(&er)

	*err = errors.New("error")
	fmt.Println("main:", *err)

	// main: error
	// defer: <nil>
}
```

**Comment by Hawken94** [comment]
了解了，原来是 defer 里的 err1 需要被后续的操作更新，明白了这个使用场景了，谢谢。
